tendermint/src/lite/verifier.rs

```diff
@@ -154,3 +165,119 @@ where
 
     Ok(())
 }
+
+/// Returns Ok if we can trust the header h2 based on h1 otherwise returns an Error.
+fn can_trust<TS, SH, VS, L, S, R>(
+    trusted_state: &TS, // h1 & h1_nextvals
+    h2: &SH,
+    trust_threshold: &L,
+    trusting_period: &Duration,
+    now: &SystemTime,
+    req: &R,
+    store: &mut S,
+) -> Result<(), Error>
+where
+    TS: TrustedState<SignedHeader = SH, ValidatorSet = VS>,
+    SH: SignedHeader,
+    VS: ValidatorSet,
+    L: TrustThreshold,
+    S: Store<SignedHeader = SH>,
+    R: Requester<SignedHeader = SH, ValidatorSet = VS>,
+{
+    match check_support(trusted_state, h2, trust_threshold, trusting_period, now) {
+        Ok(_) => {
+            store.add(h2)?;
+            return Ok(());
+        }
+        Err(e) => {
+            if e != Error::InsufficientVotingPower {
+                return Err(e);
+            }
+        }
+    }
+
+    let h1 = trusted_state.last_signed_header();
+    let h1_height: u64 = h1.header().height().value();
+    let h2_height: u64 = h2.header().height().value();
+
+    let pivot: u64 = (h1_height + h2_height) / 2;
+    let pivot_header = req.signed_header(pivot)?;
+    let pivot_next_vals = req.validator_set(pivot + 1)?;
+
+    verify(&pivot_header, &pivot_next_vals)?;
+
+    can_trust(
+        trusted_state,
+        &pivot_header,
+        trust_threshold,
+        trusting_period,
+        now,
+        req,
+        store,
+    )?;
+    store.add(&pivot_header)?;
+    let pivot_trusted = TS::new(pivot_header, pivot_next_vals);
+    can_trust(
+        &pivot_trusted,
+        h2,
+        trust_threshold,
+        trusting_period,
+        now,
+        req,
+        store,
+    )?;
+    store.add(h2)?;
+
+    Ok(())
+}
+
+/// This function captures the high level logic of the light client verification, i.e.,
+/// an application call to the light client module to (optionally download) and
+/// verify a header for some height.
+pub fn verify_header<TS, SH, VS, L, S, R>(
+    height: Height,
+    trust_threshold: &L,
+    trusting_period: &Duration,
+    now: &SystemTime,
+    req: &R,
+    store: &mut S,
+) -> Result<(), Error>
+where
+    TS: TrustedState<SignedHeader = SH, ValidatorSet = VS>,
+    L: TrustThreshold,
+    S: Store<SignedHeader = SH>,
+    R: Requester<SignedHeader = SH, ValidatorSet = VS>,
+    VS: ValidatorSet,
+    SH: SignedHeader,
+{
+    if let Ok(sh2) = store.get(height) {
+        is_within_trust_period(sh2.header(), trusting_period, now)?
+    }
+
+    let sh2 = req.signed_header(height)?;
+    let sh2_next_vals = req.validator_set(height.increment())?;
+    verify(&sh2, &sh2_next_vals)?;
+    is_within_trust_period(sh2.header(), trusting_period, now)?;
+
+    // get the highest trusted headers lower than sh2
```

Comment:
```suggestion
    // Get the highest trusted header that is lower than sh2.
```

Reply:
Changed to: Get the highest trusted header with height lower than sh2's.